### PR TITLE
Fixes #166

### LINKED ISSUE #166: Folders client needs ability to rename folder
No rush, I may implement myself if I find I need it. Just tracking a missing feature.

### PR DESCRIPTION
**Pull request by erasche**
Implements support for:

- creating folder
- updating folder
- getting folder permissions
- setting folder permissions

### CODE REVIEW

File: bioblend/galaxy/folders/__init__.py [hunk 1]
```diff
@@ -40,3 +61,74 @@ def delete_folder(self, folder_id, undelete=False):
...
+        :rtype: dict
+        :return: dictionary including details of the folder
+        """
+        url = self.gi._make_url(self, folder_id)
+        url = '/'.join([url, 'permissions'])
+        return self._get(url=url)
+
+    def set_permissions(self, folder_id, action='set_permissions', add_ids=None, manage_ids=None, modify_ids=None):
+        """
+        Display information about a folder.
```
Comment (bgruening) [inline bioblend/galaxy/folders/__init__.py]: Set folder permissions.

Comment (erasche) [inline bioblend/galaxy/folders/__init__.py]: Fixed.

**Comment by nsoranzo** [comment]
@erasche Not strictly necessary, but it would be nice to have tests for these methods, if you wish. Actually, the whole module needs tests, as `bioblend/_tests/TestGalaxyFolders.py` is sadly missing.

**Comment by erasche** [comment]
Sure, I'll try adding this. Agreed, tests would be nice.

File: bioblend/_tests/TestGalaxyFolders.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        # Then we'll add a role
+        role = self.gi.roles.get_roles()[0]
+        self.gi.folders.set_permissions(self.folder['id'], add_ids=[role['id']])
+        self.assertTrue(
+            role['id'] in
+            self.gi.folders.get_permissions(self.folder['id'], scope='available')['add_library_item_role_list'][0]
+        )
+
+        # self.folder = self.gi.folders.(self.folder['id'], self.name, 'new-description')
+        # self.assertEqual(self.folder['description'], 'new-description')
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyFolders.py]: I guess you can remove these commented lines.

File: bioblend/_tests/TestGalaxyFolders.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+"""
+Use ``nose`` to run these unit tests.
+"""
+from . import GalaxyTestBase, test_util
+
+FOO_DATA = 'foo\nbar\n'
+
+
+@test_util.skip_unless_galaxy("release_16.01")
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyFolders.py]: Can we be more granular with the test skipping? I think that you can remove this line and add the following method decorators:
- for `test_delete_folder()`: release_15.01
- for `test_update_folder()`: release_15.10
- for `test_get_set_permissions()`: release_16.01

File: bioblend/_tests/TestGalaxyFolders.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        self.assertEqual(self.folder['name'], self.name)
+        self.assertEqual(self.folder['description'], 'automatically created folder')
+
+    def test_show_folder(self):
+        f2 = self.gi.folders.show_folder(self.folder['id'])
+        self.assertEqual(f2['id'], self.folder['id'])
+
+    def test_delete_folder(self):
+        self.sub_folder = self.gi.folders.create_folder(self.folder['id'], self.name)
+        self.gi.folders.delete_folder(self.folder['id'])
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyFolders.py]: I think you want to delete `sub_folder` not `folder`.

File: bioblend/_tests/TestGalaxyFolders.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    def test_update_folder(self):
+        self.folder = self.gi.folders.update_folder(self.folder['id'], self.name, 'new-description')
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyFolders.py]: I'd also change (and check) the folder name.

**Comment by nsoranzo** [comment]
Thanks @erasche!

**Pull request merged**